framework/src/base/NonlinearSystem.C

```diff
@@ -1786,16 +1795,65 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
       {
         _fe_problem.reinitNodeFace(node, boundary_id, 0);
```

Comment:
I think it is guaranteed.  `MooseMesh::getBoundaryNodeRange()` returns iterators into the `MooseMesh::_bnd_nodes` vector.  The `_bnd_nodes` vector is created by calling   `getMesh().get_boundary_info().build_node_list(nodes, ids)`, which only adds nodes which have BCs on them.